### PR TITLE
Commitlog: Handle disk usage and disk footprint discrepancies, ensuring we flush when needed

### LINKED ISSUE #8270: The LWT stress is stopped during the prepare phase
### Installation Details
Kernel version: `5.11.3-1.el7.elrepo.x86_64`
Scylla version (or git commit hash): `4.5.dev-0.20210307.2d9feaace with build-id d779669835c747cc193d97d16b2e84c7cf05177a`
Cluster size: 4 nodes (i3.2xlarge)
OS (RHEL/CentOS/Ubuntu/AWS AMI): `ami-0291a3585729c627b` (aws: eu-north-1)

Test: `longevity-lwt-3h-test`
Test name: `longevity_lwt_test.LWTLongevityTest.test_lwt_longevity`
Test config file(s):
- [longevity-lwt-basic-3h.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/bddb74c85d77db6901179aa9b244456cf7aca521/test-cases/longevity/longevity-lwt-basic-3h.yaml)

### Issue Description ###
====================================
The test failed because of the following errors:
```
2021-03-08 04:58:37.000: (DatabaseLogEvent Severity.ERROR): type=DATABASE_ERROR regex=Exception  line_number=51623 node=Node longevity-lwt-3h-master-db-node-7e933d87-2 [13.49.21.145 | 10.0.2.141] (seed: False)
2021-03-08T04:58:37+00:00  longevity-lwt-3h-master-db-node-7e933d87-2 !INFO    | scylla: [shard 7] rpc - client 10.0.2.99:56439 msg_id 510415:  exception "Operation timed out for system.paxos - received only 0 responses from 1 CL=ONE." in no_wait handler ignored
2021-03-08 04:58:29.000: (DatabaseLogEvent Severity.ERROR): type=DATABASE_ERROR regex=Exception  line_number=26965 node=Node longevity-lwt-3h-master-db-node-7e933d87-4 [13.49.75.237 | 10.0.1.67] (seed: False)
2021-03-08T04:58:29+00:00  longevity-lwt-3h-master-db-node-7e933d87-4 !ERR     | scylla: [shard 7] paxos - CAS[164690] prune: failed exceptions::mutation_write_timeout_exception (Operation timed out for system.paxos - received only 0 responses from 1 CL=ONE.)
```

The last time the test was passed was on `07-Feb-2021 03:21:36` ([JOB#186](https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-lwt-3h-test/186/console))

All commits that were merged between the runs [JOB#186](https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-lwt-3h-test/186) - [JOB#187](https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-lwt-3h-test/187)
```
03:18 roy@Roees-MacBook-Pro ~/git/scylla/scylla ( (master)) $ git lg e1261d10f..d8345c67d
* d8345c67d - Consolidate system and non system keyspace creation (4 weeks ago) <Gleb Natapov>
* 51037e94e - lwt: handle an error during prune operation (4 weeks ago) <Gleb Natapov>
*   3dd9c5596 - Merge 'Minor tweaks to the failure detector interface' from Piotr Sarna (4 weeks ago) <Tomasz Grabiec>
|\
| * faca59efa - failure_detector: add getting last update time point (4 weeks ago) <Piotr Sarna>
| * d23584c8f - failure_detector: return arrival samples by const reference (5 weeks ago) <Piotr Sarna>
| * 445e6e44f - failure_detector: remove unimplemented is_alive method (5 weeks ago) <Piotr Sarna>
*   4acc6fecf - Merge 'locator: Check DC names in NetworkTopologyStrategy' from Juliusz Stasiewicz (4 weeks ago) <Piotr Sarna>
|\
| * 97bb15b2f - tests: Adjusted tests for DC checking in NTS (4 weeks ago) <Juliusz Stasiewicz>
| * b6fb5ee91 - locator: Check DC names in NTS (4 weeks ago) <Juliusz Stasiewicz>
* 3d001b558 - query: use local limit for non-limited queries in mixed cluster (4 weeks ago) <Botond Dénes>
* 2f50bf202 - Update seastar submodule (4 weeks ago) <Avi Kivity>
* 37b41d776 - test: add missing #include <fstream> (4 weeks ago) <Avi Kivity>
* c16e4a042 - migration_manager: Propagate schema changes with reads like we do on writes (4 weeks ago) <Tomasz Grabiec>
*   4082f57ed - Merge 'Make commitlog disk limit a hard limit.' from Calle Wilund (4 weeks ago) <Avi Kivity>
|\
| * 4be718ebf - commitlog: Force earlier cycle/flush iff segment reserve is empty (9 weeks ago) <Calle Wilund>
| * be8c359a6 - commitlog: Make segment allocation wait iff disk usage > max (9 weeks ago) <Calle Wilund>
| * ab55a1b4e - commitlog: Do partial (memtable) flushing based on threshold (9 weeks ago) <Calle Wilund>
| * 7c84b16cd - commitlog: Make flush threshold configurable (9 weeks ago) <Calle Wilund>
| * c3d95811d - table: Add a flush RP mark to table, and shortcut if not above (9 weeks ago) <Calle Wilund>
* af2d1fa0d - Update abseil submodule (5 weeks ago) <Avi Kivity>
* b9a5aff7a - distributed_loader: drop execute_futures function (5 weeks ago) <Gleb Natapov>
* a05adb853 - database: Remove global storage proxy reference (5 weeks ago) <Pavel Emelyanov>
* 8490c9ff6 - transport: Remove global storage service reference (5 weeks ago) <Pavel Emelyanov>
* 4498bb0a4 - API: Fix aggregation in column_familiy (5 weeks ago) <Amnon Heiman>
```
Apparently, the test failed due to the commit  `51037e94e - lwt: handle an error during prune operation (4 weeks ago) <Gleb Natapov>`.
@gleb-cloudius  Can you check it out?

====================================

Restore Monitor Stack command: `$ hydra investigate show-monitor 7e933d87-b67b-440f-b144-343ae967a978`
Show all stored logs command: `$ hydra investigate show-logs 7e933d87-b67b-440f-b144-343ae967a978`

Test id: `7e933d87-b67b-440f-b144-343ae967a978`

### Logs ###
**grafana** - [https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_060437/grafana-screenshot-longevity-lwt-3h-test-scylla-per-server-metrics-nemesis-20210308_060822-longevity-lwt-3h-master-monitor-node-7e933d87-1.png](https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_060437/grafana-screenshot-longevity-lwt-3h-test-scylla-per-server-metrics-nemesis-20210308_060822-longevity-lwt-3h-master-monitor-node-7e933d87-1.png)
**grafana** - [https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_060437/grafana-screenshot-overview-20210308_060437-longevity-lwt-3h-master-monitor-node-7e933d87-1.png](https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_060437/grafana-screenshot-overview-20210308_060437-longevity-lwt-3h-master-monitor-node-7e933d87-1.png)
**db-cluster** - [https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_061303/db-cluster-7e933d87.zip](https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_061303/db-cluster-7e933d87.zip)
**loader-set** - [https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_061303/loader-set-7e933d87.zip](https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_061303/loader-set-7e933d87.zip)
**monitor-set** - [https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_061303/monitor-set-7e933d87.zip](https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_061303/monitor-set-7e933d87.zip)
**sct-runner** - [https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_061303/sct-runner-7e933d87.zip](https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_061303/sct-runner-7e933d87.zip)

### Links
* [Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-lwt-3h-test/197/)
* [Download "Per server metrics nemesis" Grafana Screenshot](https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_060437/grafana-screenshot-overview-20210308_060437-longevity-lwt-3h-master-monitor-node-7e933d87-1.png)
* [Download "Overview metrics" Grafana Screenshot](https://cloudius-jenkins-test.s3.amazonaws.com/7e933d87-b67b-440f-b144-343ae967a978/20210308_060437/grafana-screenshot-longevity-lwt-3h-test-scylla-per-server-metrics-nemesis-20210308_060822-longevity-lwt-3h-master-monitor-node-7e933d87-1.pngt)

### Snapshots
image

image

### PR DESCRIPTION
**Pull request by elcallio**
Fixes #8270

If we have an allocation pattern where we leave large parts of segments "wasted" (typically because the segment has empty space, but cannot hold the mutation being added), we can have a disk usage that is below threshold, yet still get a disk _footprint_ that is over limit causing new segment allocation to stall. 

We need to take a few things into account: 
1.) Need to include wasted space in the threshold check. Whether or not disk is actually used does not matter here. 
2.) If we stall a segment alloc, we should just flush immediately. No point in waiting for the timer task. 
3.) Need to adjust the thresholds a bit. Depending on sizes, we should probably consider start flushing once we've used up space enough to be in the last available segment, so a new one is hopefully available by the time we hit the limit. 

Also fix edge case (for tests), when we have too few segment to have an active one (i.e. need flush everything).

### CODE REVIEW

File: test/boost/commitlog_test.cc [hunk 1]
```diff
@@ -894,3 +894,57 @@ SEASTAR_TEST_CASE(test_commitlog_shutdown_during_wait) {
...
+    try {
+        while (!done) {
+            auto now = timeout_clock::now();
+            rp_handle h = co_await with_timeout(now + 30s, log.add_mutation(uuid, size, db::commitlog::force_sync::no, [&](db::commitlog::output& dst) {
+                dst.fill('1', size);
+            }));
+            rps.put(std::move(h));
+        }
+    } catch (timed_out_error&) {
+        BOOST_FAIL("log write timed out. maybe it is deadlocked... Will not free log. ASAN errors and leaks will follow...");
```
Comment (avikivity) [inline test/boost/commitlog_test.cc:942]: Does it fail now? If so please move it to the end so someone bisecting will not have to understand the new failure.

File: db/commitlog/commitlog.cc [hunk 2]
```diff
@@ -1143,7 +1143,9 @@ db::commitlog::segment_manager::segment_manager(config c)
     // our threshold for trying to force a flush. needs heristics, for now max - segment_size/2.
     , disk_usage_threshold(cfg.commitlog_flush_threshold_in_mb.has_value() 
         ? size_t(std::ceil(*cfg.commitlog_flush_threshold_in_mb / double(smp::count))) * 1024 * 1024 
-        : (max_disk_size - (max_disk_size > (max_size/2) ? (max_size/2) : 0)))
+        : (max_disk_size -
+            (max_disk_size >= (max_size*2) ? max_size
+                : (max_disk_size > (max_size/2) ? (max_size/2) : max_disk_size/3))))
```
Comment (avikivity) [inline db/commitlog/commitlog.cc:1157]: I have to say I'm dazed by the calculation.

I also think it should be a percentage of the commitlog size, not a segment. There's no reason to expect correlation between memtable flush time and a segment size. We should probably flush at around 75% full. I don't mind doing that later though.

File: db/commitlog/commitlog.cc [hunk 3]
```diff
@@ -1336,6 +1345,10 @@ void db::commitlog::segment_manager::create_counters(const sstring& metrics_cate
                        sm::description("Holds a size of disk space in bytes used for data so far. "
                                        "A too high value indicates that we have some bottleneck in the writing to sstables path.")),
 
+        sm::make_gauge("disk_slack_bytes", totals.wasted_size_on_disk,
+                       sm::description("Holds a size of disk space in bytes of allocated but unusable data so far. "
+                                       "A too high value indicates that we do not write enough data to each segment.")),
```
Comment (avikivity) [inline db/commitlog/commitlog.cc:1350]: There's too kinds of slack: padding for block boundary, and slack at the end of the segment. IIUC this is slack at the end. Please clarify it in the description here.

File: db/commitlog/commitlog.cc [hunk 4]
```diff
@@ -1529,6 +1529,10 @@ future<db::commitlog::segment_manager::sseg_ptr> db::commitlog::segment_manager:
     if (!cfg.allow_going_over_size_limit && max_disk_size != 0 && totals.total_size_on_disk >= max_disk_size) {
         clogger.debug("Disk usage ({} MB) exceeds maximum ({} MB) - allocation will wait...", totals.total_size_on_disk/(1024*1024), max_disk_size/(1024*1024));
         auto f = cfg.reuse_segments ? _recycled_segments.not_empty() :  _disk_deletions.get_shared_future();
+        if (!f.available()) {
+            _new_counter = 0; // zero this so timer task does not duplicate the below flush
+            flush_segments(0); // force memtable flush already
+        }
         return f.handle_exception([this](auto ep) {
```
Comment (avikivity) [inline db/commitlog/commitlog.cc:1537]: I think we should avoid a flush if we've requested one already and it isn't complete (or is just partially complete, having flushed a few memtables), so that we don't create lots of small sstables. Do we have a notification about when a flush ends, or do we just get replay position releases?

**Comment by avikivity** [comment]
Looks good, some comments/questions.

File: test/boost/commitlog_test.cc (hunk 1 above)
Comment (elcallio) [reply test/boost/commitlog_test.cc:942]: It does. I will.

File: db/commitlog/commitlog.cc (hunk 2 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:1157]: Agree. This is just trying to make it something like "we are in the last segment".

File: db/commitlog/commitlog.cc (hunk 4 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:1537]: This is a bit risky, because we can't be sure that a flush call _will_ actually result in stuff really being released. In the case of memtables, we should probably be safe, since at most we should eventually get _more_ stuff released than we asked for. 
We could maybe replace the _new_counter check/set here and on_timer with simply having a slightly fuzzy "flush active" bool, which resets at first callback that actually frees data (since we cannot know how many there will be). Maybe per cf_id...

Comment (elcallio) [reply db/commitlog/commitlog.cc:1537]: No, it will not work. The problem is that memtables can flush themselves without us asking for it. So we could be notified of data being released before we add a bookkeep var, and then never receive the flush we asked for since memtable is already past the RP. So we'd end up thinking stuff is pending endlessly. 

I think keeping it like it is above is safest + good enough. We will only get here when we've exhausted segments from replenisher, which means nothing will enter this code until we've at least satisfies releasing one segment. Since this typically will release _more_ than   one segment, and we should not exhaust next segment until a "little later", we should not hit any congestion here.

File: test/boost/commitlog_test.cc (hunk 1 above)
Comment (elcallio) [reply test/boost/commitlog_test.cc:942]: Done

File: db/commitlog/commitlog.cc (hunk 3 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:1350]: Done

File: db/commitlog/commitlog.cc (hunk 4 above)
Comment (avikivity) [reply db/commitlog/commitlog.cc:1537]: Ok. It could be improved, but I think we need to simplify commitlog before we complexify it even more.

File: db/commitlog/commitlog.cc (hunk 2 above)
Comment (bhalevy) [reply db/commitlog/commitlog.cc:1157]: The threshold calculation doesn't make sense to me.

If `max_disk_size >= (max_size*2)` the threshold will be at 50% or higher of max_disk_size.
But if `(max_size/2) < max_disk_size < (max_size*2)` the threshold could be very close to 100% if max_disk_size is only slightly larger than `max_size/2`.  Did you mean to compare to `max_size` instead? Like this:

```
        : (max_disk_size -
            (max_disk_size >= (max_size*2) ? max_size
                : (max_disk_size > max_size ? (max_size/2) : max_disk_size/3))))
```

File: db/commitlog/commitlog.cc [hunk 5]
```diff
@@ -1874,10 +1895,11 @@ void db::commitlog::segment_manager::on_timer() {
         // above threshold, request flush.
         if (_new_counter > 0) {
             auto max = disk_usage_threshold;
-            auto cur = totals.active_size_on_disk;
+            auto cur = totals.active_size_on_disk + totals.wasted_size_on_disk;
+
             if (max != 0 && cur >= max) {
-                _new_counter = 0;
                 clogger.debug("Used size on disk {} MB exceeds local threshold {} MB", cur / (1024 * 1024), max / (1024 * 1024));
+                _new_counter = 0;
```
Comment (bhalevy) [inline db/commitlog/commitlog.cc:1902]: why the change in the placement of `_new_counter = 0`?

**Review by bhalevy — CHANGES_REQUESTED** [review]

**Comment by bhalevy** [comment]
I this PR is queued already...
I had a question/suggestion regarding the new threshold calculation:

If `max_disk_size >= (max_size*2)` the threshold will be at 50% or higher of max_disk_size.
But if `(max_size/2) < max_disk_size < (max_size*2)` the threshold could be very close to 100% if max_disk_size is only slightly larger than `max_size/2`.  Did you mean to compare to `max_size` instead? Like this:

```
        : (max_disk_size -
            (max_disk_size >= (max_size*2) ? max_size
                : (max_disk_size > max_size ? (max_size/2) : max_disk_size/3))))
```

File: db/commitlog/commitlog.cc (hunk 5 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:1902]: spurious diff.

**Comment by elcallio** [comment]
* If max_disk_size is greater than 2*segment_size (max_size), we set the threshold to (max_disk_size - max_size) - i.e. when we've used up all but the last segment.
* If max_disk_size is less than 2'segment_size
  * if max_disk_size is still larger than segment size, we set the threshold to when ~half a segment is left.
  * if max_disk_size is less or equal to segment_size (wtf!?) we set it to 2/3 load.

**Comment by elcallio** [comment]
And yes, if the size is ludicrously small, the threshold will be very close to the top.

**Comment by bhalevy** [comment]
Okay, I see where you were going now.

**Comment by avikivity** [comment]
Was reverted in d3e5b370591f443124c9cb1b304348869555cf71, caused commitlog_test timeouts in debug mode.

**Comment by elcallio** [comment]
Hm? Worked here me thinks. Will investigate.

**Comment by elcallio** [comment]
Still works for me. Not fast by any means, but completes. Will try rebasing and see if it makes any difference.

**Comment by elcallio** [comment]
Nope. Works. Do you have a reference to the test failure? Is the timeout simply too low? On my machine it taskes ~2:30m.

**Comment by elcallio** [comment]
Hm, did not get seastar module updated/synced properly. Shall rebuild again.

**Comment by elcallio** [comment]
Ok, got the rebase done properly finally (glö). Can repro now.

**Comment by avikivity** [comment]
What does glö mean? Google translate refuses to translate it, from which I deduce it is very bad.

**Comment by elcallio** [comment]
It is an expression of exasperation. Swedish version of "gah" or "ugh".
